### DISCUSSION TITLE
Random branchPrefix/additionalBranchPrefix

### DISCUSSION THREAD

**Discussion by Ziemnior** [renovatebot/renovate#15610] [Request Help]
Hello. I've been trying to force Renovate to use randomly generated string as part of branch name, in my case - `additionalBranchPrefix`. What I've tried was to write js script that on CI loads `default.json`, appends `additionalBranchPrefix` to it and then saves it. It didn't work, because preset changed only in the docker container, but not on repository `master` branch. Currently is there a way to achieve this?

My reasoning behind this is as follows: I have multiple repos which are being tracked by Renovate. As a last step of a CI pipeline, docker image is being built and pushed to the docker registry with some tag. Then, in another repo (integraton tests), branches with the same name as those created by Renovate, are being created, with updated docker images tags. Everything is okay unitl now, but after merging those MRs created by Renovate, branches in the another repo are not being updated until the next Renovate run (but not all of them!). If for example, version of one dependency used only in one of the repos change, then integration test will start with docker image tags from the last time, except this one from the service that Renovate created an MR for. Such tests may result in a wrong result, as said service can be commits ahead, break api along the way, etc. 

Randomly generated string, used as a part of a branch name would address this - after every Renovate run, if there is a new version of some dependency, then completely new branch in the integration tests repo will be created with latest (meaning: same as on master branch) docker image tags and services tags, that were affected by the dependency update.I'm self-hosting btw. 

Even though i'm no ts dev, I think I'd be able to contribute if you see an added value in such parameter/change to the existing one.
Piotr

**Comment by rarkins** [comment]
I'm not sure how to do this, but perhaps a new handlebars helper to generate a random string added here: https://github.com/renovatebot/renovate/blob/main/lib/util/template/index.ts

**Reply by rarkins** [reply]
Actually there's a potential blocker to that. Branch names need to be consistent between runs, otherwise you'll keep getting duplicates. I don't see how that's compatible with randomness.

**Reply by Ziemnior** [reply]
How about introducing something like `randomSuffix` and if such parameter is set to true, while checking if branch exists already, drop the suffix? I understand that this would require quite some rework around `branchExists`, among others.

**Reply by rarkins** [reply]
No, this would add too much complexity/burden to the code base.

**Reply by Ziemnior** [reply]
Understandable, I agree that such change would have huge impact on the codebase and further development.

**Reply by rarkins** [reply]
Do you need truly random branch names, or just to ensure that you never have the same branch name shared between different repos?

**Reply by Ziemnior** [reply]
I'd like to introduce some randomness to branch names, so every new MR would create unique branch per one run, but still would be able to detect if there is an open MR with deps version bumps. 
So for example, new version of python requests library was released -> Renovate in repositories A, B and C creates branch named `update-deps-<some random string>`. 
Two days later, new version of flask was released -> Renovate in repositories A and C creates branch named `update-deps-<different random string>`. 
On a third day, again new version of flask was released, but previous MR was not merged yet -> Renovate in repositories A and C goes to branch `update-deps-<same random string as day before>` and updates flask version to X.Y.Z+1.

I think I'll be able to mitigate this obstacle with some workaround on my side, it won't be perfect, but it will work (most of the time?). But before doing this, I wanted to ask if there's a way to do that already (as I don't know whole Renovate codebase) or if such feature could be implemented at rather small cost.

**Reply by viceice** [reply]
Why you don't use the normal renovate branch names? So you get

1. `requets` 1.1.0 -> 1.2.0: `renovate/requests-1.x`
1. `flask ` 1.0.1 -> 1.0.2: `renovate/flask-1.x`
1. `requets` 1.1.0 -> 2.0.0: `renovate/requests-2.x`
1. `requets` 1.1.0 -> 1.2.1: `renovate/requests-1.x`

**Reply by viceice** [reply]
you can also set `branchPrefix=update-deps-`, so you get 

1. `update-deps-requests-1.x`
1. `update-deps-flask-1.x`
1. `update-deps-requests-2.x`

https://docs.renovatebot.com/configuration-options/#branchprefix

**Reply by Ziemnior** [reply]
Och, I should've explain how I've configured Renovate: in the default preset, I've created two groups: external and internal and I've added apropriate rules to distinguish between them. So if anything thats is present on pypi will be updated, then it falls into external bucket and version changes will be commited to branch `renovate-external`. 

What I've figured out is to rebase branch `renovate-external` in the integration tests repository after every commit, so it will always have  service versions same as on master branch, except those modified with this commit coming from other repos. Because this diff between branch `renovate-external` and `master` was an issue that I've wanted to adress in the first place and the reason why I'd like to have some sort of randomness in branch names. This way, Renovate will behave exactly as I would like to. Thanks for the discussion though.

**Reply by viceice** [reply]
Please write down your full workflow, i don't understand it. 

Renovate can handle multiple `baseBranches` too, so if you need to update two branches with renovate, that works out of the box and never has branch collisions.

It's still not clear why you need random branch name parts.

**Reply by Ziemnior** [reply]
I'll try to elaborate a little bit more.
I have 4 repositories: service A, service B, package C and Integration Tests. Package C is a package developed internally and utilized by both services A and B. Whenever new version is released, this pacakge version is being updated by renovate - since it's an internal package, branch name is `renovate-internal`. Now, in both service repositories, branch named `renovate-internal` is being created AND as a last step in a gitlab pipeline, new branch named `renovate-internal` is being created in integration test repository, followed by two commits, as each service update its version with sed (which pipeline creates this branch doesn't matter, whichever pipeline is first). The problem is that branch that is being created as a last step of the pipeline, inherits its name from service branch name. So, if I would create a branch named `test123` in service A, then after pipeline succeeds, there will be a `test123` branch in integration tests repository. 
So, if renovate would append some random suffix to branch names per one run, then instead of `renovate-internal`, in both services branch named something like `renovate-internal-6jf5is` would be created and branch with same name would also be created in integration tests repository. Such feature would help mitigate issue with some services not being up-to-date - for example, if theres `renovate-internal` branch created at 1st of June and on 21st of June new version of package C would be released, then on said branch only services that utilize this package would have a newest version - if for example there's service D, which in turn does not use package C, then its version wouldn't change. Such behaviour could potentially lead to false positives etc in the integration tests.

But looking at the codebase of renovate and having in mind @rarkins remarks on that, I agree that such feature would require a lot of work and would have probably limited added value for majority of users. There's a problem on my side, not renovate and because we've decided to go with multirepo, now the problem is ours to solve. Having that in mind, I've decided that the best way to mitigate issue with having outdated service versions is to delete branch `renovate-internal` from integration tests repository after merging all MRs related to updating package versions. Even though it still requires a litte of mouse work, it's a milestone, not having to make multiple MRs by hand.
